internal/aws/cwlogs/pusher.go

```diff
@@ -330,17 +296,127 @@ func (p *logPusher) addLogEvent(logEvent *Event) *eventBatch {
 }
 
 func (p *logPusher) renewEventBatch() *eventBatch {
-	p.batchUpdateLock.Lock()
-	defer p.batchUpdateLock.Unlock()
 
 	var prevBatch *eventBatch
 	if len(p.logEventBatch.putLogEventsInput.LogEvents) > 0 {
 		prevBatch = p.logEventBatch
-		p.logEventBatch = newEventBatch(PusherKey{
+		p.logEventBatch = newEventBatch(StreamKey{
 			LogGroupName:  *p.logGroupName,
 			LogStreamName: *p.logStreamName,
 		})
 	}
 
 	return prevBatch
 }
+
+// A Pusher that is able to send events to multiple streams.
+type multiStreamPusher struct {
+	logStreamManager LogStreamManager
+	client           Client
+	pusherMap        map[StreamKey]Pusher
+	logger           *zap.Logger
+}
+
+func newMultiStreamPusher(logStreamManager LogStreamManager, client Client, logger *zap.Logger) *multiStreamPusher {
+	return &multiStreamPusher{
+		logStreamManager: logStreamManager,
+		client:           client,
+		logger:           logger,
+		pusherMap:        make(map[StreamKey]Pusher),
+	}
+}
+
+func (m *multiStreamPusher) AddLogEntry(event *Event) error {
+	if err := m.logStreamManager.InitStream(event.StreamKey); err != nil {
+		return err
+	}
+
+	var pusher Pusher
+
+	if _, ok := m.pusherMap[event.StreamKey]; !ok {
+		pusher = NewPusher(event.StreamKey, 1, m.client, m.logger)
+		m.pusherMap[event.StreamKey] = pusher
+	} else {
+		pusher = m.pusherMap[event.StreamKey]
+	}
```

Comment:
for this to work, it would have to be something like:
```
        var ok bool
	if pusher, ok = m.pusherMap[event.StreamKey]; !ok {
			pusher = NewPusher(event.StreamKey, 1, m.client, m.logger)
			m.pusherMap[event.StreamKey] = pusher
		}
```

Reply:
The only reason I could see to do this would be to avoid doing the lookup twice. I may be trying to over optimize here so I'll just leave this conversation open in case someone else has an opinion.

Reply:
I ended up implementing your suggestion! thanks!